pkg/controller/plan/adapter/openstack/builder.go

```diff
@@ -61,17 +210,203 @@ func (r *Builder) VirtualMachine(vmRef ref.Ref, object *cnv.VirtualMachineSpec,
 	return
 }
 
+func (r *Builder) mapFirmware(vm *model.Workload, object *cnv.VirtualMachineSpec) {
+	var firmwareType string
+	var bootloader *cnv.Bootloader
+	if imageFirmwareType, ok := vm.Image.Properties[FirmwareType]; ok {
+		firmwareType = imageFirmwareType.(string)
+	} else {
+		for _, volume := range vm.Volumes {
+			if volume.Bootable == "true" {
+				if volumeFirmwareType, ok := volume.VolumeImageMetadata[FirmwareType]; ok {
+					firmwareType = volumeFirmwareType
+				}
+			}
+		}
+	}
+	switch firmwareType {
+	case EFI:
+		// We disable secure boot even if it was enabled on the source because the guest OS won't
+		// be able to boot without getting the NVRAM data. So we start the VM without secure boot
+		// to ease the procedure users need to do in order to make the guest OS to boot.
+		secureBootEnabled := false
+		bootloader = &cnv.Bootloader{
+			EFI: &cnv.EFI{
+				SecureBoot: &secureBootEnabled,
+			}}
+	default:
+		bootloader = &cnv.Bootloader{BIOS: &cnv.BIOS{}}
+	}
+	features := &cnv.Features{}
+	firmware := &cnv.Firmware{}
+	firmware.Bootloader = bootloader
+	object.Template.Spec.Domain.Features = features
+	object.Template.Spec.Domain.Firmware = firmware
+}
+
+func (r *Builder) mapVideo(vm *model.Workload, object *cnv.VirtualMachineSpec) {
+	videoModel := DefaultProperties[VideoModel]
+	if imageVideoModel, ok := vm.Image.Properties[VideoModel]; ok {
+		videoModel = imageVideoModel.(string)
+	}
+	switch videoModel {
+	case None:
+	default:
+		autoAttachGraphicsDevice := false
+		object.Template.Spec.Domain.Devices.AutoattachGraphicsDevice = &autoAttachGraphicsDevice
+	}
```

Comment:
don't we miss setting the video model?

Reply:
AFAIK Kubevirt API does not support setting the video model, the only way I [found](https://groups.google.com/g/kubevirt-dev/c/pEEjTIARJZc?pli=1) to do it is by creating a hook sidecar: https://github.com/kubevirt/kubevirt/tree/main/cmd/example-hook-sidecar.

Reply:
ack, I see we don't have a 'validation' for the video type when the source provider is of type 'ovirt' so I think that's ok to just override it